### PR TITLE
[WIP] feat(codelabs): add intermediate lit-html codelab

### PR DESCRIPTION
**Pull request by LarsDenBakker**
Still WIP

### CODE REVIEW

File: docs/faq/lit-element-lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  /**
+   * Called when the element is removed from the page. Useful for cleaning up work
+   * done in connectedCallback.
+   */
+  disconnectedCallback() {
+    super.disconnectedCallback();
+  }
+
+  /**
+   * Called when an update was trigger, before rendering. Receives a Map of changed
```
Comment (thepassle) [inline docs/faq/lit-element-lifecycle.md]:
```suggestion
   * Called when an update was triggered, before rendering. Receives a Map of changed
```

File: docs/faq/lit-element-lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+   * properties, and their previous values. This can be used for modifying or setting
+   * new properties before a render occurs.
+   */
+  update(changed) {
+    super.update();
+  }
+
+  /**
+   * Called when an update was trigger, after rendering. Receives a Map of changed
```
Comment (thepassle) [inline docs/faq/lit-element-lifecycle.md]:
```suggestion
   * Called when an update was triggered, after rendering. Receives a Map of changed
```

File: docs/faq/lit-element-lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+   * render here yet.
+   *
+   * Can be called multiple times when your element is moved around on the page,
+   * for example when using drag & drop.
+   *
+   * You might need to make sure your setup work runs only once, or you can clean
+   * it up in the disconnectedCallback.
+   */
+  connectedCallback() {
+    super.connectedCallback();
```
Comment (thepassle) [inline docs/faq/lit-element-lifecycle.md]: Maybe a note on _why_ we have to call super.connectedcallback?

File: packages/codelabs/src/intermediate/lit-html.md [hunk 4]
```diff
@@ -0,0 +1,1363 @@
...
+</details>
+
+## Google news API
+
+For this codelab, we will be fetching live data from the google news API. You will need to retrieve an API key first:
+
+- Go to https://newsapi.org/
+- Click 'Get API key'
+- Register for an account.
+- **tip**: you don't need to use an actual valid e-mail address
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]: nice 😄

File: packages/codelabs/src/intermediate/lit-html.md [hunk 5]
```diff
@@ -0,0 +1,1363 @@
...
+      this.fetchArticles();
+    }
+  }
+}
+```
+
+</details>
+
+<details>
+<summary>Example render articles as JSON</summary>
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]:
```suggestion
<summary>Example of articles rendered as JSON</summary>
```

File: packages/codelabs/src/intermediate/lit-html.md [hunk 6]
```diff
@@ -0,0 +1,1363 @@
...
+---
+
+<details>
+ <summary>View final result</summary>
+
+```html
+<!DOCTYPE html>
+<html>
+  <body>
+    <todo-app></todo-app>
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]:
```suggestion
    <news-app></news-app>
```

File: packages/codelabs/src/intermediate/lit-html.md [hunk 7]
```diff
@@ -0,0 +1,1363 @@
...
+```
+
+</details>
+
+<details>
+<summary>Display a list of articles</summary>
+
+```js
+render() {
+ return html`
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]: indentation

File: packages/codelabs/src/intermediate/lit-html.md [hunk 8]
```diff
@@ -0,0 +1,1363 @@
...
+
+Any serious news app allows the user to track whether or not they have read an article, so let's add this capability to ours as well.
+
+As a start, you can maintain a local property for the read/unread status in the `news-article` component. We will look into lifting this data to the parent component in the next step.
+
+### Tasks to complete this step
+
+- Add a property on the `news-article` component which indicates whether the user has read the article.
+- Display the read/unread status in the title of each article.
+- Add a button on the `news-article` component to toggle between the read/unread status, storing this status locally.
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]:
```suggestion
- Add a button in the `news-article` component to toggle between the read/unread status, storing this status locally.
```

File: packages/codelabs/src/intermediate/lit-html.md [hunk 9]
```diff
@@ -0,0 +1,1363 @@
...
+
+```js
+render() {
+ return html`
+ <h3>${this.title} (${this.read ? 'read' : 'unread'})</h3>
+ `;
+}
+```
+
+If the logic is a bit more complex, separating this into a pure function is very useful. The advantage here is that we can use regular if statement, so we don't need to squash everything into a single expression:
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]:
```suggestion
If the logic is a bit more complex, separating this into a pure function is very useful. The advantage here is that we can use regular if statements, so we don't need to squash everything into a single expression:
```

File: packages/codelabs/src/intermediate/lit-html.md [hunk 10]
```diff
@@ -0,0 +1,1363 @@
...
+    `;
+  }
+}
+```
+
+</details>
+
+<details>
+ <summary>Adding event listeners</summary>
+You can add event listeners using the `@` syntax. It is just sugar for `addEventListener`:
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]:
```suggestion
With lit-html, you can add event listeners using the `@` syntax, which is just syntax sugar for `addEventListener`:
```

File: packages/codelabs/src/intermediate/lit-html.md [hunk 11]
```diff
@@ -0,0 +1,1363 @@
...
+
+```js
+render() {
+ return html`
+ <button @click=${this._onClick}></button>
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]:
```suggestion
   <button @click=${this._onClick}></button>
```

File: packages/codelabs/src/intermediate/lit-html.md [hunk 12]
```diff
@@ -0,0 +1,1363 @@
...
+</html>
+```
+
+</details>
+
+## Add a read/unread counter
+
+Now that the user can mark articles as read/unread, we want to display the total amount of read and unread items in the app. This counter should be displayed in the `news-app`, but we're storing the read/unread status in the `news-article` component. We need to think of a better way to solve this...
+
+It's best to keep the data in your application flowing in one direction from to bottom. Parent components are responsible for data of child components, including changing this data.
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]:
```suggestion
It's best to keep the data in your application flowing in one direction from top to bottom. Parent components are responsible for data of child components, including changing this data.
```

File: packages/codelabs/src/intermediate/lit-html.md [hunk 13]
```diff
@@ -0,0 +1,1363 @@
...
+
+</details>
+
+<details>
+ <summary>Update read status in main app</summary>
+
+To update the read status, we need to use immutable data update patterns. This means we should create a articles array, and a new object for the article that was updated. A quick way to do this, is by using a map function:
+
+```js
+_toggleReadStatus(articleToUpdate) {
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]: indentation

File: packages/codelabs/src/intermediate/lit-html.md [hunk 14]
```diff
@@ -0,0 +1,1363 @@
...
+ <summary>View final result</summary>
+
+```html
+<!DOCTYPE html>
+<html>
+  <body>
+    <todo-app></todo-app>
+
+    <script type="module">
+      import { LitElement } from 'https://unpkg.com/lit-element?module';
```
Comment (thepassle) [inline packages/codelabs/src/intermediate/lit-html.md]:
```suggestion
      import { LitElement, html } from 'https://unpkg.com/lit-element?module';
```

File: docs/faq/lit-element-lifecycle.md (hunk 3 above)
Comment (LarsDenBakker) [reply docs/faq/lit-element-lifecycle.md]: Good point, added it

**Pull request merged**